platforms/core-configuration/base-diagnostics/src/main/java/org/gradle/api/tasks/diagnostics/TaskReportTask.java

```diff
@@ -245,44 +252,63 @@ private List<RuleDetails> ruleDetailsFor(Project project) {
 
     private DefaultGroupTaskReportModel taskReportModelFor(Project project, boolean detail) {
         final AggregateMultiProjectTaskReportModel aggregateModel = new AggregateMultiProjectTaskReportModel(!detail, detail, getDisplayGroup(), getDisplayGroups());
-        final TaskDetailsFactory taskDetailsFactory = new TaskDetailsFactory(project);
 
-        final SingleProjectTaskReportModel projectTaskModel = buildTaskReportModelFor(taskDetailsFactory, project);
-        aggregateModel.add(projectTaskModel);
+        ProjectIdentity relativeProjectIdentity = ((ProjectInternal) project).getProjectIdentity();
 
-        for (final Project subproject : project.getSubprojects()) {
-            aggregateModel.add(buildTaskReportModelFor(taskDetailsFactory, subproject));
-        }
+        Stream.concat(Stream.of(project), project.getSubprojects().stream())
+            .map(p -> buildTaskReportModelFor(relativeProjectIdentity, p))
+            .forEach(aggregateModel::add);
 
         aggregateModel.build();
 
         return DefaultGroupTaskReportModel.of(aggregateModel);
     }
 
-    private SingleProjectTaskReportModel buildTaskReportModelFor(
-        final TaskDetailsFactory taskDetailsFactory,
-        final Project subproject
-    ) {
-        return projectStateFor(subproject).fromMutableState(
-            project -> SingleProjectTaskReportModel.forTasks(
-                getProjectTaskLister().listProjectTasks(project),
-                taskDetailsFactory
-            )
-        );
+    private SingleProjectTaskReportModel buildTaskReportModelFor(ProjectIdentity relativeProjectIdentity, Project p) {
+        return ((ProjectInternal) p).getOwner().fromMutableState(project -> {
+            ImmutableSetMultimap.Builder<String, TaskDetails> groups = ImmutableSetMultimap.<String, TaskDetails>builder()
+                .orderKeysBy(String::compareToIgnoreCase)
+                .orderValuesBy(Comparator.comparing(TaskDetails::getPath));
+
+            for (Task task : getProjectTaskLister().listProjectTasks(project)) {
+                groups.put(
+                    getGroupFor(task),
+                    TaskDetails.of(getPathFor(relativeProjectIdentity, task), task)
+                );
+            }
+
+            return new SingleProjectTaskReportModel(groups.build());
+        });
     }
 
-    private ProjectState projectStateFor(Project subproject) {
-        return getProjectStateRegistry().stateFor(subproject);
+    /**
+     * Determines the path of this task, relative to the given project ID.
+     */
+    private static Path getPathFor(ProjectIdentity relativeProjectIdentity, Task task) {
+        TaskIdentity<?> taskId = ((TaskInternal) task).getTaskIdentity();
+        ProjectIdentity taskProjectIdentity = taskId.getProjectIdentity();
+
+        boolean isParentProject = taskProjectIdentity.getBuildTreePath()
+            .startsWith(relativeProjectIdentity.getBuildTreePath());
```

Comment:
💅 This feels like it should be a method on `ProjectIdentity`.

Reply:
Good idea. Added `isParentOf` and `isChildOf` methods to `ProjectIdentity` + unit tests